### PR TITLE
feat: add deposit service

### PR DESCRIPTION
**Pull request by npty**
# Description

This PR added support for wrap/unwrap deposit address generation.

## Changes

- Added `getNativeWrapDepositAddress` API
- Added `getNativeUnwrapDepositAddress` API
- Renamed `getDepositAddress` to `getLinkedDepositAddress`
- Added the new `getDepositAddress` API. This function will return the deposit address based on three scenarios:
   - The source token is native token, then this function returns deposit address from the `getNativeWrapDepositAddress` function.
   - The source token is not native token and the associated `asset` on the destination chain has `isERC20WrappedNativeGasToken = true`, then this function will return the linked deposit address where the destination address for linked deposit address is unwrap deposit address. (there're two addresses generation)
   - The source token is not native token and `isERC20WrappedNativeGasToken` is falsy, then the function simply returns the linked deposit address
   
 ## Todo
 - [x] ~~Validate the wrap/unwrap addresses offline to prevent man in the middle attack~~ Skip
 - [x] Validate if given chains are active
 - [x] Manually check on testnet if those addresses from 3 cases are working properly.
 
### 1. Native Wrap Address

Sending AVAX from Avalanche to Fantom should receive WAVAX at the destination chain

#### Parameters
```ts
{
  sourceChain: "Avalanche",
  destinationChain: "Fantom",
  destinationAddress: "0xB8Cd93C83A974649D76B1c19f311f639e62272BC"
}
```

#### Returned Address

`0xcd780367774c6ff9d781f41b09a095c450ecda58`

#### Test Result
Passed ✅

[Avalanche Tx](https://testnet.snowtrace.io/tx/0x970835a0b6d99a2d37a10958fa9918b1e6855ecfcff221ca97420efcfa39b1f7)
[Axelarscan Tx](https://testnet.axelarscan.io/transfer/0x988c19c553cdd31c1e4cd9496c6d4309e7c30ea36fa673a4713a6c04714b87ac)
[Fantom Tx](https://testnet.ftmscan.com/tx/0x3d568ed93b96cbb4d8eac16562128d43cf615d364e3a232d95ce295bf2aa8a7c)

### 2. Unwrap Address

Sending WFTM from Avalanche to Fantom should receive FTM at the destination chain

#### Parameters
```ts
{
     sourceChain: "Avalanche",
     destinationChain: "Fantom",
     asset: "wftm-wei",
     destinationAddress: "0xB8Cd93C83A974649D76B1c19f311f639e62272BC",
}
```

#### Returned Address

`0x7C74188B53193E2513C707F26d1Ecf6d0d2765B4`

#### Test Results

Passed ✅

[Avalanche Tx](https://testnet.snowtrace.io/tx/0xe729fe538b4a8afbd621f4a4bfd0d2abe01779e708cffd8c7dae6ee6ff73f635?chainId=43113)
Axelarscan Tx not found
[Fantom Tx](https://testnet.ftmscan.com/tx/0x955e5c91daa424b5990ad6d55e19cdd657207f77aa085ce84af9f9820cc7cad6)

### 3. Simple Linked Address

Sending aUSDC from Avalanche to Fantom should receive aUSDC at the destination chain

#### Parameters
```ts
{
    sourceChain: "Avalanche",
    destinationChain: "Fantom",
    asset: "uasdc",
    destinationAddress: "0xB8Cd93C83A974649D76B1c19f311f639e62272BC"
}
```

#### Returned Address

`0xE12F77E36745Ad1019fbd19d09f2491047f9Ee73`

#### Test Result

Passed ✅

[Avalanche Tx](https://testnet.snowtrace.io/tx/0x0fe8b77414c5e296ab9b90763dc2a8a1ce0a3a5d76ede716e15f4057704b285c?chainId=43113)
Axelarscan Tx not found
[Fantom Tx](https://testnet.ftmscan.com/tx/0x17a0e9af1c49ea0769c819b3a6507a241ce94ea84121f58ca8928815e988d226)

### CODE REVIEW

File: packages/deposit-address/src/get-deposit-address/isomorphic.ts [hunk 1]
```diff
@@ -1,16 +1,104 @@
...
+        destinationChain,
+        environment,
+        salt: params.options?.salt,
+        refundAddress: params.options?.refundAddress || destinationAddress,
+      },
+      dependencies
+    );
+  } else {
+    // this is an erc20 token, we need to check if it is unwrappable at the destination chain
+    const shouldUnwrapToken = unwrappable(
```
Comment (canhtrinh) [inline packages/deposit-address/src/get-deposit-address/isomorphic.ts:55]: we should still give the user the option of receiving the erc20 on the destination chain, i.e. not always unwrapping

File: packages/deposit-address/src/get-deposit-address/isomorphic.ts [hunk 2]
```diff
@@ -32,3 +120,66 @@ export async function getDepositAddress(
...
+
+export async function getNativeWrapDepositAddress(
+  params: DepositNativeWrapOptions,
+  dependencies: GetDepositServiceDependencies
+) {
+  const chainConfigs = await dependencies.configClient.getChainConfigs(
+    params.environment
+  );
+
+  validateChainIds([params.sourceChain, params.destinationChain], chainConfigs);
```
Comment (canhtrinh) [inline packages/deposit-address/src/get-deposit-address/isomorphic.ts:132]: with all these validations, can we externalize this to a helper function?

File: packages/deposit-address/src/get-deposit-address/isomorphic.ts [hunk 3]
```diff
@@ -32,3 +120,66 @@ export async function getDepositAddress(
...
+
+export async function getNativeUnwrapDepositAddress(
+  params: DepositNativeUnwrapOptions,
+  dependencies: GetDepositServiceDependencies
+) {
+  const chainConfigs = await dependencies.configClient.getChainConfigs(
+    params.environment
+  );
+
+  validateChainIds([params.sourceChain, params.destinationChain], chainConfigs);
```
Comment (canhtrinh) [inline packages/deposit-address/src/get-deposit-address/isomorphic.ts:164]: with all these validations, can we externalize this to a helper function?

File: packages/deposit-address/src/get-deposit-address/isomorphic.ts (hunk 2 above)
Comment (npty) [reply packages/deposit-address/src/get-deposit-address/isomorphic.ts:131]: yeah, this function is already exported here https://github.com/axelarnetwork/axelarjs/blob/833eff620eec627254213e85c4e7489403ab0a15/packages/deposit-address/src/get-deposit-address/helpers/validation.ts#L30

File: packages/deposit-address/src/get-deposit-address/types.ts [hunk 4]
```diff
@@ -14,9 +15,44 @@ export type SendOptions = {
   environment: Environment;
 };
 
-export type GetDepositAddressDependencies = {
+export type DepositAddressOptions = {
+  sourceChain: string;
+  destinationChain: string;
+  destinationAddress: string;
+  environment: Environment;
+  asset?: string; // if not specified, it is the native asset
```
Comment (canhtrinh) [inline packages/deposit-address/src/get-deposit-address/types.ts:23]: how about `string | 'native';`?

File: packages/deposit-address/src/get-deposit-address/isomorphic.ts (hunk 2 above)
Comment (npty) [reply packages/deposit-address/src/get-deposit-address/isomorphic.ts:131]: these functions are already put under [helpers/validation.ts](https://github.com/axelarnetwork/axelarjs/blob/833eff620eec627254213e85c4e7489403ab0a15/packages/deposit-address/src/get-deposit-address/helpers/validation.ts#L30)

or did you mean putting all of these validations into a one big helper function?

```ts
  validateChainIds([params.sourceChain, params.destinationChain], chainConfigs);
  validateAsset(
    [params.sourceChain, params.destinationChain],
    params.asset,
    chainConfigs
  );
  validateAddress(
    params.destinationAddress,
    chainConfigs.chains[params.destinationChain.toLowerCase()] as ChainConfig
  );
 ```

File: packages/deposit-address/src/get-deposit-address/types.ts (hunk 4 above)
Comment (npty) [reply packages/deposit-address/src/get-deposit-address/types.ts:23]: this doesn't work since "native" is overridden by string

Comment (npty) [reply packages/deposit-address/src/get-deposit-address/types.ts:23]: how about we make `asset` to be required, not optional, but remove `getNativeWrapDepositAddress` from the `getDepositAddress` function?

So if people want to send native token, they have to use `getNativeWrapDepositAddress` instead.

**Review by alanrsoares — APPROVED** [review]

**Review by canhtrinh — APPROVED** [review]

**Pull request merged**